### ISSUE TITLE
Should non-OSS licenses and associated source be stripped by prep-source-build.sh?

### ISSUE THREAD

**Issue by MichaelSimons** [dotnet/source-build#4736]
This issue stems from a discussion in https://github.com/dotnet/sdk/pull/44828.

The VMR contains source to build the Windows installers.  In some cases there are non-OSS licenses included in the VMR for these installers.  Currently these licenses have to be special cased when scanning the source for non-OSS licenses - e.g. https://github.com/dotnet/sdk/blob/262b9c3d6cf67287f649e38d83e6c5d9d08feb8a/src/SourceBuild/content/test/Microsoft.DotNet.SourceBuild.Tests/assets/LicenseScanTests/LicenseExclusions.txt#L144.  This adds a tax for each distro maintainer who scans the source before building around analyzing and ensuring the source is not part of the source built product.  To provide a better UX, should the prep-source-build.sh strip these licenses and the associated source similar to how binaries are stripped today?

**Comment by MichaelSimons** [comment]
@dotnet/distro-maintainers - Do you have thoughts on this?  Would this be helpful to you?

**Comment by mateusrodrigues** [comment]
Yes, I believe this would be helpful.

We do run our own source scan before every package upload and analyze such cases that non-free licenses are found, e.g. #4590. If the script is able to strip those beforehand, that's even better.

And similar to the binary removal tool, one could simply skip that step if they want to.

**Comment by omajid** [comment]
> To provide a better UX, should the prep-source-build.sh strip these licenses and the associated source similar to how binaries are stripped today?

In Fedora, we have two conflicting goals when it comes to building .NET from source:

- We want to use the pristine upstream sources, and validate that they haven't been tampered with
- We don't want to ever include non-OSS sources or binaries

To address the first goal, we use the github tarball, which contains the full source of the VMR, with nothing added or removed.

To address the second goal, we scan *everything* in the source tarball. This happens in a context where nothing (like prep-source-build.sh) has executed, and we are only working with inputs to the build system.

So this change, which strips up things after extracting from the original/unmodified/pristine tarball, doesn't really affect us.If it helps others users, then +1 for making this.

**Comment by ellahathaway** [comment]
After discussing this offline with @MichaelSimons and @mthalman, I think a few changes will need to be made to the license scan infrastructure in order for the license removal tool to function appropriately. Below is a proposal for the changes to the license scanner and implementation details for the license removal tool:

- License Scanner as a VMR Test:
  - The license scanner is considered to be a VMR test, but it currently exists in `test/Microsoft.DotNet.SourceBuild.Tests` instead of `test/Microsoft.DotNet.Tests`. We should move the scanner to the appropriate test project.
  - Since the license scan test is a VMR test and not a source-build test, we should not run `./prep-source-build.sh` before running the license scanner. This means that the stripping of non-SB allowed licenses will not be part of the license scan test.

- Exclusions Baselines:
  - The LicenseExclusions.txt baseline should be removed and broken up into two new files:
      - allowed-vmr-licenses.txt
      - allowed-sb-licenses.txt
   - Any licenses not listed in allowed-sb-licenses.txt will be stripped by a tool that removes files from the VMR
   - Any licenses not included in allowed-vmr-licenses will be detected as "new" licenses. This approach is similar to the current binary tooling process.
   - Below is a diagram of the proposed exclusion baselines:

![Image](https://github.com/user-attachments/assets/61f2180a-5d9a-4483-a070-c949b98039b4)

License Removal Tool Functionality:
- The removal tool will not detect licenses. This functionality is part of the license scanner test. It will only remove non-sb allowed licenses.
- It will be runnable via prep-source-build.sh to remove non-SB allowed licenses.
- detect-binaries.sh should be removed in favor of creating a single script that invokes a vmr tools. This script will be called `vmr-tool.sh`
   - `vmr-tool.sh` will act similarly to `eng/common/sdk-tasks.ps1`
      - It will prep and run a standalone tool in the VMR under `eng/tools` (but not in `eng/tools/tasks`)
      - It will accept --tool, --with-packages, and --with-sdk as input, and will filter additional args to the tool itself.

License Scan Test Functionality:
- Should be updated to consume the new baselines: allowed-vmr-licenses.txt and allowed-sb-licenses.txt
- Should be updated to remove unused exclusions from both baselines, like the BinaryToolKit does.

**Comment by MichaelSimons** [comment]
Thanks for the detailed proposal @ellahathaway.  I have a couple questions/comments.

1. Will the allowed allowed-vmr-licenses.txt reference the allowed-sb-licences.txt?  I want to make sure we are not duplicating allows.  1. You don't explicitly mention it but I assume the allowed files should also be moved outside of the test infra.  Putting them next to the allowed-*-binaries.txt files feels like a good location.
1. Moving the license scanner outside of the SB tests is covered in https://github.com/dotnet/source-build/issues/4253 and feels out of scope for this work.
1. How will we validate the removal tool is working correctly and doesn't regress?  Will we be running the a version of the license scanner against the prepped source?

**Comment by ellahathaway** [comment]
> 1. Will the allowed allowed-vmr-licenses.txt reference the allowed-sb-licences.txt?  I want to make sure we are not duplicating allows. 

Yes, I was thinking the allowed-vmr-licenses.txt will reference the allowed-sb-licenses.txt with a line like `import:allowed-sb-licenses.txt`

> 1. You don't explicitly mention it but I assume the allowed files should also be moved outside of the test infra.  Putting them next to the allowed-*-binaries.txt files feels like a good location.

Yes, I was planning on moving these files. Thanks for calling this out.

> 1. Moving the license scanner outside of the SB tests is covered in [Move license scanning out of the SB smoke tests #4253](https://github.com/dotnet/source-build/issues/4253) and feels out of scope for this work.

This sounds good to me.

> 1. How will we validate the removal tool is working correctly and doesn't regress?  Will we be running the a version of the license scanner against the prepped source?

Since the license scanner will run against `allowed-vmr-licenses.txt`, both baselines (vmr and sb allowed licenses) will be updated by the `CreateBaselineUpdatePRTool`. This tooling will handle keeping the baselines that the file remover uses up to date.

As far as functionality of the tool, I will write unit tests to ensure that the tool maintains functionality and does not regress.
